// @stamtech/datatable-utils/esm2015/lib/components/dialog-filter/dialog-filter.component.js
import { Component, Inject } from '@angular/core';
import { FormControl, FormGroup } from '@angular/forms';
import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
export class DialogFilterComponent {
    constructor(dialog, data) {
        this.dialog = dialog;
        this.data = data;
        this.filtersForm = new FormGroup({});
        this.filtersService = data.filtersService;
        this.filtersService.filterSet$.subscribe(filterSet => {
            this.filterSet = filterSet;
            Object.keys(this.filterSet.filters).forEach(key => this.filtersForm.addControl(key, new FormControl({ value: this.filterSet.filters[key].value, disabled: this.filterSet.filters[key].disabled })));
        });
    }
    clearFilters() {
        this.filtersForm.reset();
    }
    search() {
        Object.keys(this.filtersForm.controls).forEach(control => {
            const filter = this.filterSet.find(control);
            if (filter)
                filter.value = this.filtersForm.controls[control].value;
        });
        this.filterSet.resetPage();
        this.filtersService.updateFilters(this.filterSet);
        this.dialog.close();
    }
}
DialogFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-dialog-filter',
                template: "<h2 mat-dialog-title>Filtri</h2>\r\n<div mat-dialog-content>\r\n    <form [formGroup]=\"filtersForm\">        \r\n        <div *ngFor=\"let filter of filterSet.filters | keyvalue\">\r\n            <ng-container [ngSwitch]=\"filter.value.type\">\r\n                <ng-container *ngSwitchCase=\"'TEXT'\">\r\n                    <mat-form-field>\r\n                        <mat-label>{{ filter.value.label }}</mat-label>\r\n                        <input matInput [formControlName]=\"filter.key\">\r\n                    </mat-form-field>\r\n                </ng-container>\r\n                <ng-container *ngSwitchCase=\"'SELECT'\">\r\n                    <mat-form-field>\r\n                        <mat-label>{{ filter.value.label }}</mat-label>\r\n                        <mat-select [formControlName]=\"filter.key\">\r\n                            <mat-option [value]=\"\">---</mat-option> \r\n                            <mat-option *ngFor=\"let opt of filter.value.options\" [value]=\"opt.id\">\r\n                                {{ opt.label }}\r\n                            </mat-option>\r\n                        </mat-select>\r\n                    </mat-form-field>\r\n                </ng-container>\r\n                <ng-container *ngSwitchCase=\"'DATE'\">\r\n                    <mat-form-field>\r\n                        <mat-label>{{ filter.value.label }}</mat-label>\r\n                        <input matInput [matDatepicker]=\"picker\" [formControlName]=\"filter.key\">\r\n                        <mat-datepicker-toggle matSuffix [for]=\"picker\"></mat-datepicker-toggle>\r\n                        <mat-datepicker #picker></mat-datepicker>\r\n                    </mat-form-field>\r\n                </ng-container>\r\n            </ng-container>\r\n        </div>\r\n    </form>\r\n</div>\r\n<mat-dialog-actions align=\"end\">\r\n    <button class=\"clear-filters-button\" mat-raised-button (click)=\"clearFilters()\">Reset</button>\r\n    <button class=\"search-filters-button\" mat-raised-button color=\"primary\" (click)=\"search()\">Ricerca</button>\r\n</mat-dialog-actions>",
                styles: [".mat-form-field { width: 100% }"]
            },] }
];
DialogFilterComponent.ctorParameters = () => [
    { type: MatDialogRef },
    { type: undefined, decorators: [{ type: Inject, args: [MAT_DIALOG_DATA,] }] }
];
