// @stamtech/datatable-utils/esm2015/lib/services/columns-manager.service.js
import { DatatableColumn } from "../shared/datatable.column";
export class ColumnsManagerService {
    constructor() {
        this.columns = {};
    }
    addColumn(key, label) {
        this.columns[key] = new DatatableColumn(key, label);
    }
    setVisible(keys, value = true) { keys.forEach(k => { this.columns[k].visible = value; }); }
    setLocked(keys, value = true) { keys.forEach(k => { this.columns[k].locked = value; }); }
    setCustom(keys, value = true) { keys.forEach(k => { this.columns[k].customContent = value; }); }
    setSortable(keys, sortKey = "") { keys.forEach(k => { this.columns[k].sortable = sortKey; }); }
    // Get all columns
    getColumns() {
        return Object.values(this.columns);
    }
    // Get only visible columns
    getDisplayedColumns() {
        return Object.values(this.columns).filter(c => c.visible).map(c => c.key);
    }
    // Get only standard columns ( the content is simply the value )
    getStandardColumns() {
        return Object.values(this.columns).filter(c => !c.customContent);
    }
}
