// @stamtech/datatable-utils/esm2015/lib/services/column-filter.service.js
import { BehaviorSubject } from 'rxjs';
import { Filter, FilterSet } from '../shared/datatable.types';
export class ColumnFilterService {
    constructor() {
        this.filterSet = new FilterSet();
        this.filterSetSubject = new BehaviorSubject(this.filterSet);
        this.filterSet$ = this.filterSetSubject.asObservable();
    }
    emitFilters() { this.filterSetSubject.next(this.filterSet); }
    // Filters methods  
    addFilter(key, label, type, options) {
        this.filterSet.add(new Filter(key, label, type, options));
        this.emitFilters();
    }
    removeFilter(key) {
        this.filterSet.remove(key);
        this.emitFilters();
    }
    setFilter(key, value) {
        let filter = this.filterSet.find(key);
        if (filter) {
            filter.value = value;
            this.emitFilters();
        }
    }
    clearFilter(key) { this.setFilter(key, null); }
    updateFilters(filterSet) {
        this.filterSet = filterSet;
        this.emitFilters();
    }
    isActive() {
        return this.filterSet.isActive();
    }
    disableFilter(key) {
        let filter = this.filterSet.find(key);
        if (filter) {
            filter.disabled = true;
            this.emitFilters();
        }
    }
    enableFilter(key) {
        let filter = this.filterSet.find(key);
        if (filter) {
            filter.disabled = false;
            this.emitFilters();
        }
    }
    clearFilters() {
        this.filterSet.clear();
        this.emitFilters();
    }
    // Paging methods
    setPage(page, limit) {
        this.filterSet.page = page;
        if (limit)
            this.filterSet.limit = limit;
        this.emitFilters();
    }
    setLimit(limit) {
        this.filterSet.limit = limit;
        this.emitFilters();
    }
    // Sorting methods
    setSort(field, direction) {
        this.filterSet.sortField = field;
        this.filterSet.sortDirection = direction;
        this.emitFilters();
    }
}
