// @stamtech/datatable-utils/esm2015/lib/shared/datatable.datasource.js
import { BehaviorSubject, of } from "rxjs";
import { catchError, finalize } from "rxjs/operators";
import { RowsResponse } from "../services/datatable.service.interface";
export class DatatableDatasource {
    constructor(rowsService) {
        this.rowsService = rowsService;
        this.rows = new BehaviorSubject([]);
        this.loadingSubject = new BehaviorSubject(false);
        this.loading$ = this.loadingSubject.asObservable();
        this.countSubject = new BehaviorSubject(0);
        this.counter$ = this.countSubject.asObservable();
    }
    connect(collectionViewer) {
        return this.rows.asObservable();
    }
    disconnect(collectionViewer) {
        this.rows.complete();
        this.loadingSubject.complete();
        this.countSubject.complete();
    }
    loadData(filterSet) {
        this.loadingSubject.next(true);
        this.rowsService.findAll(filterSet)
            .pipe(catchError(() => of(new RowsResponse())), finalize(() => this.loadingSubject.next(false)))
            .subscribe((result) => {
            this.rows.next(result.content);
            this.countSubject.next(result.totalElements);
        });
    }
}
