// @stamtech/datatable-utils/esm2015/lib/shared/datatable.types.js
import { HttpParams } from "@angular/common/http";
export class Filter {
    constructor(key, label, type, options) {
        this.value = null;
        this.disabled = false;
        this.key = key;
        this.label = label;
        this.type = type;
        if (options)
            this.options = options.map((value) => {
                if (typeof value === 'string') {
                    return { id: value, label: value };
                }
                return value;
            });
    }
}
export var FilterType;
(function (FilterType) {
    FilterType["TEXT"] = "TEXT";
    FilterType["SELECT"] = "SELECT";
    FilterType["DATE"] = "DATE";
})(FilterType || (FilterType = {}));
/**
 * Track the current datatable filter status with :
 * - an array of applicable filters with their value
 * - the current pagination status
 * - the current sort field
 */
export class FilterSet {
    constructor() {
        // Applicable filters
        this.filters = {};
        // Pagination
        this.limit = 10;
        this.page = null; // Zero based
        // Sorting
        this.sortField = null;
        this.sortDirection = 'Desc';
    }
    find(key) { return this.filters[key]; }
    add(filter) { this.filters[filter.key] = filter; }
    remove(key) { delete this.filters[key]; }
    isActive() { return this.getActiveFilters().length > 0; }
    clear() {
        Object.keys(this.filters).forEach(key => {
            this.filters[key].value = null;
        });
    }
    reset() { this.filters = {}; }
    getActiveFilters() {
        return Object.values(this.filters).filter(f => (f.value !== null && f.value !== undefined));
    }
    getParams() {
        console.log(this);
        let params = new HttpParams();
        if (this.page !== null) {
            params = params.append('page', this.page.toString());
            params = params.append('limit', this.limit.toString());
        }
        if (this.sortField) {
            params = params.append('sort', this.sortField.toString());
            params = params.append('direction', this.sortDirection.toString());
        }
        if (this.filters) {
            this.getActiveFilters().forEach((filter) => params = params.append(filter.key, filter.value));
        }
        return params;
    }
    setSort(field, direction) {
        this.sortField = field;
        this.sortDirection = direction;
        this.resetPage();
    }
    resetPage() {
        if (this.page)
            this.page = 0;
    }
}
