// @stamtech/datatable-utils/esm2015/lib/components/column-selector/column-selector.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { moveItemInArray } from '@angular/cdk/drag-drop';
export class ColumnSelectorComponent {
    constructor() {
        this.menuVisible = false;
        this.columns = [];
        this.columnsChange = new EventEmitter();
        this.selectedColumns = [];
        this.selectedColumnsChange = new EventEmitter();
    }
    ngOnInit() { }
    toggleSelector() {
        this.menuVisible = !this.menuVisible;
    }
    // On backdrop click close the menu and send selectedColumnsChange event
    closeMenu() {
        this.menuVisible = false;
        this.selectedColumnsChange.emit(this.selectedColumns);
    }
    // After drop the array is reordered
    drop(event) {
        const previousColumn = this.columns[event.previousIndex];
        const currentColumn = this.columns[event.currentIndex];
        moveItemInArray(this.columns, event.previousIndex, event.currentIndex);
        const previousSelectIndex = this.selectedColumns.indexOf(previousColumn.key);
        const currentSelectIndex = this.selectedColumns.indexOf(currentColumn.key);
        if (previousSelectIndex != -1 && currentSelectIndex != -1) {
            moveItemInArray(this.selectedColumns, previousSelectIndex, currentSelectIndex);
        }
    }
}
ColumnSelectorComponent.decorators = [
    { type: Component, args: [{
                selector: 'column-selector',
                template: "<button\r\n    mat-icon-button\r\n    color=\"primary\"\r\n    (click)=\"toggleSelector()\"\r\n    cdkOverlayOrigin #trigger=\"cdkOverlayOrigin\">\r\n    <mat-icon aria-label=\"Columns\">view_column</mat-icon>\r\n</button>\r\n<ng-template\r\n  cdkConnectedOverlay\r\n  [cdkConnectedOverlayOrigin]=\"trigger\"\r\n  [cdkConnectedOverlayOpen]=\"menuVisible\"\r\n  [cdkConnectedOverlayHasBackdrop]=\"true\"\r\n  (backdropClick)=\"closeMenu()\">\r\n    <mat-selection-list class=\"column-list mat-elevation-z8\"\r\n        [(ngModel)]=\"selectedColumns\"\r\n        cdkDropList  (cdkDropListDropped)=\"drop($event)\">\r\n        <mat-list-option *ngFor=\"let column of columns\" \r\n            class=\"column-item\" \r\n            checkboxPosition=\"before\"  \r\n            [disabled]=\"column.locked\"\r\n            [value]=\"column.key\"\r\n            cdkDrag>\r\n            <div class=\"columnItem\">\r\n              <span>{{column.label}} </span>\r\n              <mat-icon>drag_handle</mat-icon>\r\n            </div>\r\n          </mat-list-option>\r\n    </mat-selection-list>\r\n</ng-template>\r\n",
                styles: [".column-list{min-width:200px;max-width:100%;border:1px solid #ccc;max-height:350px;display:block;background:#eee;overflow-y:auto}.cdk-drag-preview{display:flex;height:48px;-webkit-tap-highlight-color:transparent;width:100%;padding:0;position:relative}.cdk-drag-placeholder{opacity:0}.cdk-drag-animating{transition:transform .25s cubic-bezier(0,0,.2,1)}.columnItem{display:flex;justify-content:space-between}.columnItem mat-icon{cursor:-webkit-grab;cursor:grab}"]
            },] }
];
ColumnSelectorComponent.ctorParameters = () => [];
ColumnSelectorComponent.propDecorators = {
    columns: [{ type: Input }],
    columnsChange: [{ type: Output }],
    selectedColumns: [{ type: Input }],
    selectedColumnsChange: [{ type: Output }]
};
