// @stamtech/datatable-utils/esm2015/lib/components/chips-filter/chips-filter.component.js
import { Component } from '@angular/core';
import { ColumnFilterService } from '../../services/column-filter.service';
export class ChipsFilterComponent {
    constructor(filtersService) {
        this.filtersService = filtersService;
        this.chips = [];
        this.filtersService.filterSet$
            .subscribe(filterSet => this.chips = filterSet.getActiveFilters());
    }
    clear(key) { this.filtersService.clearFilter(key); }
    clearFilters() { this.filtersService.clearFilters(); }
}
ChipsFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'stam-chips-filter',
                template: "<div class=\"infobox-container\" *ngIf=\"chips.length\">\r\n    <div class=\"mat-header-cell filter-header-label\" >Filtri attivi : </div>\r\n    <mat-chip-list>\r\n        <mat-chip *ngFor=\"let chip of chips\" \r\n            color=\"primary\"\r\n            selected=\"true\"\r\n            (removed)=\"clear(chip.key)\">\r\n            <mat-icon matChipRemove>cancel</mat-icon>\r\n            {{chip.label}} <span class=\"chipValue\">{{chip.value}}</span>\r\n        </mat-chip>\r\n    </mat-chip-list>\r\n    <div class=\"action-container\">\r\n        <button mat-button color=\"accent\" (click)=\"clearFilters()\">Cancella filtri</button>\r\n    </div>\r\n</div>\r\n\r\n",
                styles: [".infobox-container{display:flex;align-items:center;min-height:56px;width:100%;background:#fff;border-bottom:1px solid rgba(0,0,0,.12)}.infobox-container .filter-header-label{width:8em;margin:0 24px}.infobox-container mat-chip-list{flex-grow:1;margin:1em 0}.infobox-container mat-chip-list .chipValue{background-color:hsla(0,0%,100%,.5);padding:7px 16px;border-radius:16px;margin-left:8px;margin-right:-12px;line-height:19px;max-width:8em;overflow:hidden;white-space:nowrap;text-overflow:ellipsis}.infobox-container mat-chip-list mat-icon{margin-left:0!important;margin-right:8px!important}"]
            },] }
];
ChipsFilterComponent.ctorParameters = () => [
    { type: ColumnFilterService }
];
